File: 0398-random-pick-index/0398-random-pick-index.ts
class Solution {

    private mp: Map<number, number[]>;
    constructor(nums: number[]) {
        this.mp = new Map();

        const n : number = nums.length;

        for(let i = 0; i < n; i++){
            if(!this.mp.has(nums[i])){
                this.mp.set(nums[i], [])
            }

            this.mp.get(nums[i])!.push(i);
        }
    }

    pick(target: number): number {
        const li : number[] = this.mp.get(target)!;

        const idx: number = Math.floor(Math.random() * li.length);

        return li[idx];
    }
}

File: 0398-random-pick-index/0398-random-pick-index.java
class Solution {
    Map<Integer, List<Integer>> mp;
    public Solution(int[] nums) {
        int n = nums.length;

        mp = new HashMap<>();

        for(int i = 0; i < n; i++){
            if(!mp.containsKey(nums[i])){
                mp.put(nums[i], new ArrayList<>());
            }

            mp.get(nums[i]).add(i);
        }
    }
    
    public int pick(int target) {
        List<Integer> li = mp.get(target);

        int index = (int) (Math.random() * li.size());

        return li.get(index);
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 */